crates/wit-parser/src/resolve.rs

```diff
@@ -184,13 +158,9 @@ impl Resolve {
     ///
     /// Any dependency resolution error or otherwise world-elaboration error
     /// will be returned here. If successful a package identifier is returned.
-    pub fn push(
-        &mut self,
-        mut unresolved: UnresolvedPackage,
-        deps: &HashMap<String, PackageId>,
-    ) -> Result<PackageId> {
+    pub fn push(&mut self, mut unresolved: UnresolvedPackage) -> Result<PackageId> {
```

Comment:
With the removal of `deps` here, how are registry-aware tooling supposed to pass external dependencies to the parser?

What would be nice would be a callback mechanism for "load me this external dependency" when encountered for the first time, passing it the package name and getting back a `Result<Option<UnresolvedPackage>>` (`parse_dir` could be reimplemented on top of that rather than parsing everything under `deps`).

Reply:
Sorry, I think I'm confusing myself after staring at a diff for the past few hours.

I see how this is supposed to work with the new design; as package ids are a thing now, we can simply merge the decoded worlds from the dependencies download from the registry and push the `UnresolvedPackage` for the top-level target WIT file; a mapping of external package name to arena id (i.e. the `deps` parameter here) is no longer necessary.

So I guess my problem is simply the redundant version needed to make this work, where previously it wasn't a problem:

For example, I have the following in `Cargo.toml`:

```toml
"wasi:http" = "1.0"
```

And in my WIT I have:

```wit
world foo {
   import wasi:http/foo@1.0
}
```

That's fine until a 1.1 of the `wasi:http` package is published I run `cargo component update` and now it fails to parse the WIT sans any other changes. Worse, the versions will go out of sync with these two files (but `Cargo-component.lock` would be in sync , of course).

Reply:
To an extent the versions only exist in WIT insofar as to have the ability to extract WIT from a preexisting component in a somewhat lossless way. In other words by admitting that the binary format will have versions in it we were then forced to have versions here too.

That being said the way I see this problem is "ok we're forced to have versions but how can we not?" When initially talking with @lukewagner I was also concerned about the version repetition involved. One somewhat-saving grace is that WIT isn't intended to be all that large so there shouldn't be that many versions, but your point still stands of it'd be better to not do this.

One possible half-baked idea, though, is:

* Registry tooling uses `UnresolvedPackge` to learn about the dependencies of a WIT package.
* Idiomatically no one writes down versions in their dependencies.
* Registry tooling will see you depend on `wasi:http/foo`, verify it's in the manifest, and record its version.
* Tooling adds any necessary dependencies, including transitive ones, to `Resolve` (with versions)
* Tooling edits `UnresolvedPackage::foreign_deps` to list versions and then adds the end-user WIT to a `Resolve`.

That way versions are everywhere in the internal representation but to the user they declared the version of their own package, declared the versions of dependencies, and otherwise didn't interact with versions at all. Later on when WIT is extracted from the final component it'll also have all the versions in place too.

Now whether or not this is a good idea is up for debate I think.

Reply:
I'll fully admit I haven't fully thought out all the use cases, but yeah what you outlined is what I was thinking.